### ISSUE TITLE
AnomalyGPT: Detecting Industrial Anomalies Using Large Vision-Language Models

### ISSUE THREAD

**Issue by YoojLee** [YoojLee/paper_review#86]
<img width="1118" alt="image" src="https://github.com/user-attachments/assets/8ad218e9-a6f0-4f4e-a0e9-2a52cd9cc477">

## Summary
LVLM을 Industrial anomaly detection task에 learnable prompt를 활용하여 잘 adapt하는 연구. NSA를 활용해서 anomaly를 simulation하고, 이에 맞춰 decoder 및 prompt를 학습시킴. 학습된 prompt는 llm에 continuous vector 형태로 feed됨. AnomalyGPT 같은 경우에는 visual memory를 활용하여 few shot transfer 형태로 few-shot learning 진행 (즉, few-shot learning을 진행할 때에 auxiliary dataset에서의 meta-training 과정을 한 번 거친다). 높은 성능을 보여주고 있고, 일단 주 contribution은 LVLM을 IAD에 성공적으로 적용했다는 정도?

## Method Highlights
<img width="1412" alt="image" src="https://github.com/user-attachments/assets/a574e729-3e16-4aff-8935-79b631798368">

- multi-class 방식으로 학습됨 (uniAD와 비슷한 형태). 
- decoder와 prompt learner로 크게 구성되며, decoder에서는 localization을 학습하고, prompt learner의 경우에는 decoder output이랑 이미지 임베딩을 llm token embedding space랑 align해주는 형태라고 보면 될 것 같음.
- unsupervised training을 거치고, few-shot transfer 수행.
    - few-shot transfer에서는 별도의 학습을 거치는 건 아니고, in-context learning 형태로 가는 듯함 (이때, 추가의 prompt learning을 거치는지는 아직 unclear하긴 한데 in-context learning이라고 한 데에서 학습은 안하는 게 의도된 것으로 보임)

### Decoder
unsup이랑 few-shot transfer에서 localization result $M$이 다르게 정의됨. 이렇게 만들어진 $M$은 prompt learner에 original image embedding이랑 같이 feed되어서 concat된 후, 최종 LLM에 들어가는 형태
- unsupervised learning
심플하게 layer 별로 text랑 이미지의 dot product 연산을 거쳐서 gt mask의 resolution으로 reshape & upsample되는 형태. 중간 이미지 feature는 text랑 align이 안되어 있어서 align시키는 learnable linear 레이어 추가로 통과하여 학습.

<img width="655" alt="image" src="https://github.com/user-attachments/assets/0e225abc-e0b9-45b6-b506-faebbbc951f4">

- few-shot transfer
레이어 별로 reference 이미지들과의 similarity map을 계산해서 add하는 형태로 localization pred 계산.

<img width="673" alt="image" src="https://github.com/user-attachments/assets/8c5891fd-7d56-4e0e-acaf-962bd6176eb5">

### Prompt Learner
여기서 prompt learning은, encoder 안에 inject되어서 token embedding의 representation을 forward 연산 과정에서 변형시키는 형태로 가는 게 아님. 여기에서 textual representation의 space에는 변함이 없고, 최종 output을 adjust하는 형태로 prompt learning이 진행되는 것.

<img width="947" alt="image" src="https://github.com/user-attachments/assets/f1672a38-a6ad-4314-8758-3ae7fb44005d">

- unsupervised learning 단계에서는 prompt를 scratch로 학습함.
- 그 다음에 Few shot 단계에서는 추가적인 prompt 학습 없이 inference만 진행하는 것으로 보임 (in-context learning이니까..?)
- unsup인지 few shot인지에 따라 image encoder를 통과한 embedding과 concat되는 localization result가 다름.
- llm에 태운 후에 llm의 output을 next token prediction task로 auto-regressive하게 학습하는 형태로 감.
- gt 구성은 그냥 사전 정의한 룰에 따라서 anomaly가 존재하면, anomaly 있다 + 있는 위치 매핑해주고 (그 rule은 아래 그림 참고) 기존 템플릿에 따라 만들어주는 듯. (보다 자세한 디테일은 [링크](https://github.com/CASIA-IVA-Lab/AnomalyGPT/blob/main/code/datasets/all_supervised_with_cn.py#L631) 참고) 

<img width="674" alt="image" src="https://github.com/user-attachments/assets/060fc866-153b-44f9-b3cb-6e444a2f2785">

- 참고로, text encoder에 들어가는 prompt와 llm에 들어가는 textual prompt는 별도로 존재하며, 후자의 경우에는 class specific하게 들어감.

1. text encoder에 들어가는 textual prompt (following WinCLIP)
    
<img width="1370" alt="image" src="https://github.com/user-attachments/assets/0f98ca90-42e9-4b2f-97a1-ce5b08db0d1f">

2. llm에 들어가는 prompt

<img width="679" alt="image" src="https://github.com/user-attachments/assets/e2d63b75-36fc-42ce-b441-b4aa8baec552">
<img width="1038" alt="image" src="https://github.com/user-attachments/assets/959ccd09-c87a-463f-8154-6df1d81210f3">

### 학습 관련
- ce loss (prompt learner 거쳐서 llm prediction과 gt sequence 간의 cross entropy 계산)
- dice loss & focal loss -> loss for localization (two track으로 localization map을 따로 gt map과 비교해서 계산하는 형태)

<img width="675" alt="image" src="https://github.com/user-attachments/assets/3ae7dab4-c8ac-4c3d-81fb-d58a9e5438b2">

- vlm 아키텍처는 pandaGPT 가져옴 (pandaGPT의 visual encoder는 ImageBind를 따르는 듯. ImageBind 내부에 clip과 동일한 형태로 text encoder도 구현되어  있음)

## Experiment Highlights

논문에 score는 어떻게 계산하는지 안나와있는데, 코드 보면 auroc 점수는 decoder output으로 compute + i-auroc는 localization map에서 max때리는 형태임.

<img width="1371" alt="image" src="https://github.com/user-attachments/assets/9e2425f7-40b5-4879-adff-0263b0c8e1ea">

auxiliary dataset (mvtec의 경우에는 visa, visa의 경우에는 mvtec) 에서 학습하고, 그 다음에 추가 parameter 업데이트 없이 in-context learning 형태로  few shot transfer한 성능을 report. threshold를 따로 지정하지 않고도 llm output만으로도 decision이 가능하기 때문에 accuracy도 추가로 report한 형태.

흠, 근데 few-shot inference라는 걸 확인했을 때 auroc 점수에 prompt learner가 개입하는 부분은 전혀 없음 -> 이게 과연 의미있는 점수라고 볼 수 있나? 그냥 similarity map으로 계산한 점수 아닌지?

<img width="479" alt="image" src="https://github.com/user-attachments/assets/1517eb5d-3b7f-4875-9dc7-e77f58be4303">

그냥 unsup 성능. 전체적으로 p-auroc는 좀 떨어지는 걸 알 수 있음.

<img width="985" alt="image" src="https://github.com/user-attachments/assets/bb24fd16-eb14-40f1-a48f-23e9da715ae3">

이렇게 멀티턴 대화가 가능하다고 보여주고 있음 -> 어찌 보면 당연해서 그렇게 새로운 건가 싶음
appendix에서 다른 모델과 비교한 걸 보면, pandaGPT를 이용한 다른 모델들은 어느 정도 이미지에 대한 recognition은 잘 동작하고 있음. 다만, 아래의 그림처럼 anomaly에 대한 판별 혹은 위치 특정이 잘 안됨.

<img width="650" alt="image" src="https://github.com/user-attachments/assets/8008e0e5-2108-4bad-88df-58290aba898e">

다른 vlm 아키텍처를 사용했으면 어떻게 됐을지 궁금함?

<img width="1392" alt="image" src="https://github.com/user-attachments/assets/0d0d07e8-dcd6-4d1e-beaa-2937397eaf35">

visa는 왜 1-shot transfer 성능만 보여줬을까 하는 의문이 듦 (그래서 ablation에서 전체적인 경향성 해석이 조금 까다로운 것 같음)

LoRA랑 비교한 부분이 좀 재미있었음. LoRA로 fine-tuning하면 오히려 성능이 더 떨어지더라 (lora는 참고로 accuracy만 비교하면서 보는 게 맞을 듯)  -> 없는 것보다야 lora라도 하는 게 낫지만, prompt learner를 사용하는 옵션 대비해서는 LLM의 space를 건드리지 않는 게 좋다 (이거는 근데 vlm에서도 어느 정도 비슷한 결과였던 거 같기도 하고..), pre-training보다 적은 데이터로 llm의 space를 건드리는 게 좋은 결과를 가져올 것 같진 않음.

Decoder가 prompt learning을 통해 llm의 gradient를 받게 되면 localization 성능이 더 오른다는 점은 재미있는 듯함 (p-auroc의 상승). 

## Strengths
- LVLM을 효과적으로 IAD task에 adapt
- llm을 쓰면 threshold를 굳이 찾는 과정을 겪지 않아도 되긴 해서 간편해보이긴 함

## Weaknesses
- unsup training을 거치고, few-shot transfer를 하는 거라면 같은 도메인 셋에서의 meta-training을 수행하지 않는 방법론들 (padim, patchcore, winclip) 과의 직접적인 비교는 무의미해보임.
- unified training이지만, class-specific한 description을 llm에 feed해줌으로써 class-specific한 부분을 완전히 제거할 수는 없음 + 이러한 description을 사용하지 않는 경우엔 성능이 어떻게 되는지? 여기에서 어느 정도 information leakage가 발생하지는 않을까?
- few-shot transfer에서 prompt가 개입하는 부분이 없다면 결국에 similarity map만 이용해서 auroc를 compute하는 건데 이게 무슨 의미가 있는지? (ablation 보면, visa 1-shot에서 왜 갑자기 default setting에서만 i-auroc 값이 다른지 이해가 안 감)
- instruction tuning을 하면 되는 문제를 힘들게 돌아가는 느낌. instruction tuning을 한 모델과도 비교를 해보면 재미있었을 듯함.
